src/affine_dimensions.jl

```diff
@@ -0,0 +1,452 @@
+    return AffineDimensions{promote_type(R1,R2)}
+end
+function Base.promote_rule(::Type{AffineDimensions{R1}}, ::Type{Dimensions{R2}}) where {R1,R2}
+    return Dimensions{promote_type(R1,R2)}
+end
+function Base.promote_rule(::Type{Dimensions{R1}}, ::Type{AffineDimensions{R2}}) where {R1,R2}
+    return Dimensions{promote_type(R1,R2)}
+end
+function Base.promote_rule(::Type{SymbolicDimensions{R1}}, ::Type{AffineDimensions{R2}}) where {R1,R2}
+    return Dimensions{promote_type(R1,R2)}
+end
+function Base.promote_rule(::Type{AffineDimensions{R1}}, ::Type{SymbolicDimensions{R2}}) where {R1,R2}
+    return Dimensions{promote_type(R1,R2)}
+end
+
+
+# Conversions for Dimensions |> AffineDimenions 
+"""
+    uconvert(qout::UnionAbstractQuantity{<:Any, <:AbstractAffineDimensions}, q::UnionAbstractQuantity{<:Any, <:Dimensions})
+
+Convert a quantity `q` with base SI units to the affine units of `qout`, for `q` and `qout` with compatible units.
+You can also use `|>` as a shorthand for `uconvert`
+"""
+function uconvert(qout::UnionAbstractQuantity{<:Any,<:AffineDimensions}, q::UnionAbstractQuantity{<:Any,<:Dimensions})
+    @assert isone(ustrip(qout)) "You passed a quantity with a non-unit value to uconvert."
+    dout = dimension(qout)
+    dimension(q) == basedim(dout) || throw(DimensionError(q, qout))
+    vout = (ustrip(q) - offset(dout))/scale(dout)
+    return new_quantity(typeof(q), vout, dout)
+end
+
+function uconvert(qout::UnionAbstractQuantity{<:Any,<:AffineDimensions}, q::QuantityArray{<:Any,<:Any,<:Dimensions})
+    @assert isone(ustrip(qout)) "You passed a quantity with a non-unit value to uconvert."
+    dout = dimension(qout)
+    dimension(q) == basedim(dout) || throw(DimensionError(q, qout))
+    vout = (ustrip(q) .- offset(dout))./scale(dout)
+    return QuantityArray(vout, dout, quantity_type(q))
+end
+
+function uconvert(qout::UnionAbstractQuantity{<:Any,<:AbstractSymbolicDimensions}, qin::AbstractQuantityOrArray{<:Any,<:AbstractAffineDimensions})
+    uconvert(qout, siunits(qin))
+end
+
+function uconvert(qout::UnionAbstractQuantity{<:Any,<:AbstractAffineDimensions}, qin::AbstractQuantityOrArray{<:Any,<:AbstractSymbolicDimensions})
+    uconvert(qout, siunits(qin))
+end
+
+function uconvert(qout::UnionAbstractQuantity{<:Any,<:AbstractAffineDimensions}, qin::AbstractQuantityOrArray{<:Any,<:AbstractAffineDimensions})
+    uconvert(qout, siunits(qin))
+end
+
+function map_dimensions(op::typeof(+), args::AffineDimensions...) 
+    assert_no_offset.(args)
+    return AffineDimensions(
+        scale=*(scale.(args)...),
+        offset=zero(Float64),
+        basedim=map_dimensions(op, basedim.(args)...)
+    )
+end
+
+function map_dimensions(op::typeof(-), args::AffineDimensions...) 
+    assert_no_offset.(args)
+    return AffineDimensions(
+        scale=/(scale.(args)...),
+        offset=zero(Float64),
+        basedim=map_dimensions(op, basedim.(args)...)
+    )
+end
```

Comment:
Please minimize writing the same code twice. Here, you could write this out like:

```
+for (op, combine) in ((:+, :*), (:-, :/))
+    @eval function map_dimensions(op::typeof($op), args::AffineDimensions...)
+        map(assert_no_offset, args)
+        return AffineDimensions(
+            scale=($combine)(map(scale, args)...),
+            offset=zero(Float64),
+            basedim=map_dimensions(op, map(basedim, args)...) 
+        )
+    end
 end
```

Also note the `map` instead of `.`. Using `map` tends to be safer unless you are dealing with a multi-dimensional array.